server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -269,6 +278,57 @@ public void adjustEntitlementQuantity(Consumer consumer, Entitlement ent,
         }
     }
 
+    /**
+     *
+     * This is a helper method to attempt to entitle a guest using host provided pools, if available
+     *
+     * @param data AutobindData encapsulating data required for an autobind request
+     * @param entitlements List<Entitlements> The list of entitlements to populate with entitlements,
+     *                     should this method succeed
+     * @return boolean True when a regular entitleByProducts is required.
+     *         False when this method was able to satisfy the guest and no regular entitleByProducts is necessary
+     * @throws AutobindDisabledForOwnerException when an autobind attempt is made and the owner
+     *         has it disabled.
+     */
+    private boolean entitleGuest(AutobindData data, List<Entitlement> entitlements) throws EntitlementRefusedException {
+        Consumer host;
+        List<String> hostProvidedPools;
+        Consumer consumer = data.getConsumer();
+        Owner owner = consumer.getOwner();
+
+        // We should only attempt to use the host pools if there have not already been added any pools.
+        if (!consumer.hasFact("virt.uuid") || !"true".equalsIgnoreCase(consumer.getFact("virt.is_guest"))
+            || (data.getPossiblePools() != null && !data.getPossiblePools().isEmpty())) {
+            return true;
+        }
+        host = consumerCurator.getHost(consumer.getFact("virt.uuid"), owner);
+        if (host == null) {
+            return true;
+        }
+        hostProvidedPools = poolCurator.getPoolsProvidedByHost(host);
+        if (hostProvidedPools.isEmpty()) {
+            return true;
+        }
+        data.setPossiblePools(hostProvidedPools);
+        entitlements.addAll(poolManager.entitleByProducts(data));
+        ComplianceStatus status;
+        if (data.getOnDate() != null) {
```

Comment:
null check is not needed here. getStatus takes care of the case where date is null. please also consider the other arguments eventually passed into getStatus.